File: python/meshly/snapshot.py
"""
Snapshot system for storing simulation data at different time points.

This module provides classes for saving and loading simulation snapshots
containing multiple fields of data in a compressed zip format.
"""

import json
import zipfile
from typing import Dict, Optional, List
from io import BytesIO
import numpy as np
from pydantic import BaseModel, Field

from .array import ArrayUtils, ArrayMetadata, EncodedArray
from .common import PathLike


class FieldMetadata(BaseModel):
    """Metadata for field data (without the actual array)"""
    name: str = Field(..., description="Field name")
    type: str = Field(..., description="Field type (scalar, vector, tensor)")
    units: Optional[str] = Field(None, description="Field units")
    shape: List[int] = Field(..., description="Shape of the field data array")
    dtype: str = Field(..., description="Data type of the field data array")
    itemsize: int = Field(..., description="Size of each item in bytes")


class FieldData(BaseModel):
    """General field data for simulation results"""
    name: str = Field(..., description="Field name")
    type: str = Field(..., description="Field type (scalar, vector, tensor)")
    data: np.ndarray = Field(..., description="Field data array")
    units: Optional[str] = Field(None, description="Field units")

    class Config:
        arbitrary_types_allowed = True


class SnapshotMetadata(BaseModel):
    """Metadata for a snapshot (excluding field data arrays)"""
    time: float = Field(..., description="Time value")
    fields: Dict[str, FieldMetadata] = Field(
        default_factory=dict, description="Field metadata")


class Snapshot(BaseModel):
    """
    Snapshot of simulation results at a specific time.

    A snapshot contains a time value and a dictionary of field data.
    It can be saved to and loaded from a single compressed zip file.

    Zip structure:
        metadata.json - Contains time and field metadata
        fields/
            <field_name>.bin - Encoded array data for each field
    """
    time: float = Field(..., description="Time value")
    fields: Dict[str, FieldData] = Field(
        default_factory=dict, description="Field data")

    class Config:
        arbitrary_types_allowed = True

    def get_field(self, name: str) -> Optional[FieldData]:
        """
        Get a field by name.

        Args:
            name: Field name

        Returns:
            FieldData if found, None otherwise
        """
        return self.fields.get(name)

    def add_field(
        self,
        name: str,
        data: np.ndarray,
        field_type: str = "scalar",
        units: Optional[str] = None
    ) -> None:
        """
        Add a field to the snapshot.

        Args:
            name: Field name
            data: Field data array
            field_type: Field type (scalar, vector, tensor)
            units: Optional field units
        """
        self.fields[name] = FieldData(
            name=name,
            type=field_type,
            data=data,
            units=units
        )

    @property
    def field_names(self) -> List[str]:
        """Get list of field names in this snapshot."""
        return list(self.fields.keys())


class SnapshotUtils:
    """Utility class for working with snapshots."""

    @staticmethod
    def save_to_zip(
        snapshot: Snapshot,
        destination: PathLike | BytesIO,
        date_time: Optional[tuple] = None
    ) -> None:
        """
        Save a snapshot to a zip file.

        Args:
            snapshot: Snapshot to save
            destination: Path to the output zip file or BytesIO object
            date_time: Optional date_time tuple for deterministic zip files
        """
        # Build field metadata and encode arrays
        field_metadata: Dict[str, FieldMetadata] = {}
        encoded_fields: Dict[str, bytes] = {}

        for field_name, field_data in snapshot.fields.items():
            # Encode the field array
            encoded = ArrayUtils.encode_array(field_data.data)
            encoded_fields[field_name] = encoded.data

            # Create field metadata
            field_metadata[field_name] = FieldMetadata(
                name=field_data.name,
                type=field_data.type,
                units=field_data.units,
                shape=list(encoded.shape),
                dtype=str(encoded.dtype),
                itemsize=encoded.itemsize
            )

        # Create snapshot metadata
        snapshot_metadata = SnapshotMetadata(
            time=snapshot.time,
            fields=field_metadata
        )

        # Write to zip file
        with zipfile.ZipFile(destination, "w", compression=zipfile.ZIP_DEFLATED, compresslevel=6) as zipf:
            # Prepare files to write
            files_to_write: List[tuple] = [
                ("metadata.json", json.dumps(
                    snapshot_metadata.model_dump(), indent=2, sort_keys=True))
            ]

            # Add encoded field arrays
            for field_name, encoded_data in encoded_fields.items():
                files_to_write.append(
                    (f"fields/{field_name}.bin", encoded_data))

            # Write files in sorted order for deterministic output
            for filename, data in sorted(files_to_write):
                if date_time is not None:
                    info = zipfile.ZipInfo(
                        filename=filename, date_time=date_time)
                else:
                    info = zipfile.ZipInfo(filename=filename)
                info.compress_type = zipfile.ZIP_DEFLATED
                info.external_attr = 0o644 << 16  # Fixed file permissions
                if isinstance(data, str):
                    data = data.encode('utf-8')
                zipf.writestr(info, data)

    @staticmethod
    def load_from_zip(source: PathLike | BytesIO) -> Snapshot:
        """
        Load a snapshot from a zip file.

        Args:
            source: Path to the input zip file or BytesIO object

        Returns:
            Loaded Snapshot instance
        """
        with zipfile.ZipFile(source, "r") as zipf:
            # Load metadata
            with zipf.open("metadata.json") as f:
                metadata_dict = json.loads(f.read().decode("utf-8"))
                snapshot_metadata = SnapshotMetadata(**metadata_dict)

            # Load field data
            fields: Dict[str, FieldData] = {}

            for field_name, field_meta in snapshot_metadata.fields.items():
                # Load encoded array data
                field_path = f"fields/{field_name}.bin"
                with zipf.open(field_path) as f:
                    encoded_data = f.read()

                # Create EncodedArray and decode
                encoded_array = EncodedArray(
                    data=encoded_data,
                    shape=tuple(field_meta.shape),
                    dtype=np.dtype(field_meta.dtype),
                    itemsize=field_meta.itemsize
                )
                decoded_array = ArrayUtils.decode_array(encoded_array)

                # Create FieldData
                fields[field_name] = FieldData(
                    name=field_meta.name,
                    type=field_meta.type,
                    data=decoded_array,
                    units=field_meta.units
                )

            return Snapshot(
                time=snapshot_metadata.time,
                fields=fields
            )

File: typescript/src/snapshot.ts
import JSZip from "jszip"
import { ArrayMetadata, ArrayUtils } from "./array"


/**
 * Metadata for a field (without the actual array data)
 */
export interface FieldMetadata {
    name: string
    type: string
    units: string | null
    shape: number[]
    dtype: string
    itemsize: number
}

/**
 * Metadata for a snapshot
 */
export interface SnapshotMetadata {
    time: number
    fields: Record<string, FieldMetadata>
}

/**
 * Field data with decoded array
 */
export interface FieldData {
    name: string
    type: string
    data: Float32Array | Uint32Array
    units: string | null
    shape: number[]
}

/**
 * Result of loading a snapshot
 */
export interface SnapshotResult {
    time: number
    fields: Record<string, FieldData>
}

/**
 * Utility class for loading snapshots from zip files
 */
export class SnapshotUtils {
    /**
     * Load snapshot metadata from a zip file without decoding field arrays.
     * Useful for inspecting snapshot contents before loading specific fields.
     *
     * @param zipInput The zip file as JSZip, ArrayBuffer, or Uint8Array
     * @returns Promise resolving to SnapshotMetadata
     */
    static async loadMetadata(
        zipInput: JSZip | ArrayBuffer | Uint8Array
    ): Promise<SnapshotMetadata> {
        const zip = zipInput instanceof JSZip ? zipInput : await JSZip.loadAsync(zipInput)

        // Load metadata.json
        const metadataFile = zip.file("metadata.json")
        if (!metadataFile) {
            throw new Error("metadata.json not found in snapshot zip file")
        }

        const metadataText = await metadataFile.async("text")
        return JSON.parse(metadataText) as SnapshotMetadata
    }

    /**
     * Load a specific field from a snapshot zip file.
     *
     * @param zipInput The zip file as JSZip, ArrayBuffer, or Uint8Array
     * @param fieldName The name of the field to load
     * @returns Promise resolving to FieldData
     */
    static async loadField(
        zipInput: JSZip | ArrayBuffer | Uint8Array,
        fieldName: string
    ): Promise<FieldData> {
        const zip = zipInput instanceof JSZip ? zipInput : await JSZip.loadAsync(zipInput)

        // Load metadata to get field info
        const metadata = await SnapshotUtils.loadMetadata(zip)

        const fieldMeta = metadata.fields[fieldName]
        if (!fieldMeta) {
            throw new Error(`Field "${fieldName}" not found in snapshot. Available fields: ${Object.keys(metadata.fields).join(", ")}`)
        }

        // Load the encoded field data
        const fieldPath = `fields/${fieldName}.bin`
        const fieldFile = zip.file(fieldPath)
        if (!fieldFile) {
            throw new Error(`Field data file "${fieldPath}" not found in zip`)
        }

        const encodedData = await fieldFile.async("uint8array")

        // Decode using ArrayUtils
        const arrayMetadata: ArrayMetadata = {
            shape: fieldMeta.shape,
            dtype: fieldMeta.dtype,
            itemsize: fieldMeta.itemsize
        }

        const decodedArray = ArrayUtils.decodeArray(encodedData, arrayMetadata)

        return {
            name: fieldMeta.name,
            type: fieldMeta.type,
            data: decodedArray,
            units: fieldMeta.units,
            shape: fieldMeta.shape
        }
    }

    /**
     * Load multiple fields from a snapshot zip file.
     *
     * @param zipInput The zip file as JSZip, ArrayBuffer, or Uint8Array
     * @param fieldNames Array of field names to load. If empty or undefined, loads all fields.
     * @returns Promise resolving to Record of field name to FieldData
     */
    static async loadFields(
        zipInput: JSZip | ArrayBuffer | Uint8Array,
        fieldNames?: string[]
    ): Promise<Record<string, FieldData>> {
        const zip = zipInput instanceof JSZip ? zipInput : await JSZip.loadAsync(zipInput)

        // Load metadata
        const metadata = await SnapshotUtils.loadMetadata(zip)

        // Determine which fields to load
        const fieldsToLoad = fieldNames && fieldNames.length > 0
            ? fieldNames
            : Object.keys(metadata.fields)

        // Load each field
        const fields: Record<string, FieldData> = {}
        for (const fieldName of fieldsToLoad) {
            fields[fieldName] = await SnapshotUtils.loadField(zip, fieldName)
        }

        return fields
    }

    /**
     * Load an entire snapshot from a zip file.
     *
     * @param zipInput The zip file as JSZip, ArrayBuffer, or Uint8Array
     * @returns Promise resolving to SnapshotResult containing time and all fields
     */
    static async loadFromZip(
        zipInput: JSZip | ArrayBuffer | Uint8Array
    ): Promise<SnapshotResult> {
        const zip = zipInput instanceof JSZip ? zipInput : await JSZip.loadAsync(zipInput)

        // Load metadata
        const metadata = await SnapshotUtils.loadMetadata(zip)

        // Load all fields
        const fields = await SnapshotUtils.loadFields(zip)

        return {
            time: metadata.time,
            fields
        }
    }

    /**
     * Get list of field names in a snapshot without loading the field data.
     *
     * @param zipInput The zip file as JSZip, ArrayBuffer, or Uint8Array
     * @returns Promise resolving to array of field names
     */
    static async getFieldNames(
        zipInput: JSZip | ArrayBuffer | Uint8Array
    ): Promise<string[]> {
        const metadata = await SnapshotUtils.loadMetadata(zipInput)
        return Object.keys(metadata.fields)
    }

    /**
     * Get the time value from a snapshot without loading field data.
     *
     * @param zipInput The zip file as JSZip, ArrayBuffer, or Uint8Array
     * @returns Promise resolving to the time value
     */
    static async getTime(
        zipInput: JSZip | ArrayBuffer | Uint8Array
    ): Promise<number> {
        const metadata = await SnapshotUtils.loadMetadata(zipInput)
        return metadata.time
    }
}
